### ISSUE TITLE
Support using a custom Cabal 

### ISSUE THREAD

**Issue by mboes** [commercialhaskell/stack#2910]
It's sometimes useful to use Stack with a custom version of Cabal. e.g. because a bug has been fixed upstream in `master` but not yet released on Hackage. Stack has `stack setup --upgrade-cabal`, which changes the version of Cabal used, "globally". But from what I can tell, there is no way to ask for a custom Cabal.

Something like the following would be nice:

```
$ stack setup --upgrade-cabal --cabal-sdist URL
```

Ideally though, this would only apply to the local project, not globally.

Any ideas for workarounds for this in the meantime would be appreciated.

cc @robinbb @facundominguez

**Comment by mgsloan** [comment]
Agreed that this would be handy.  See also this related issue: https://github.com/commercialhaskell/stack/issues/2386
